Taverna 3

ThilinaManamgoda opened the pull request.

ThilinaManamgoda committed:
renamed Utility class to CWLUtil

ThilinaManamgoda committed:
teting java 8 features

ThilinaManamgoda committed:
add parallel capabilities to provider

ThilinaManamgoda committed:
result is cleared

ThilinaManamgoda committed:
modified result object

ThilinaManamgoda committed:
trying port name

ThilinaManamgoda committed:
resource folder is added

ThilinaManamgoda committed:
add return statment

ThilinaManamgoda committed:
pom is edited

ThilinaManamgoda committed:
tring to run workbench

ThilinaManamgoda committed:
Merge remote-tracking branch 'apache/cwl-browse'

ThilinaManamgoda committed:
building

ThilinaManamgoda committed:
dependency confilcts resolved .
pay attention to cwlDesc getActivity class

ThilinaManamgoda committed:
dependencies resovled

ThilinaManamgoda committed:
CWLUtil test cases

ThilinaManamgoda committed:
fixed : long line in description

ThilinaManamgoda committed:
add constants

ThilinaManamgoda committed:
T3 way

ThilinaManamgoda committed:
remove snakeyaml dep from ui module

stain commented on taverna-cwl-activity-ui/pom.xml:
```diff
@@ -65,27 +60,45 @@
-			<artifactId>taverna-cwl-activity</artifactId>
-			<groupId>${project.groupId}</groupId>
-			<version>${project.version}</version>
+			<groupId>net.sf.taverna.t2.ui-api</groupId>
```
These `net.sf.taverna.*` need to be updated to `org.apache.taverna.*` Taverna 3 dependencies, e.g.:

``` xml
        <dependency>
            <groupId>org.apache.taverna.workbench</groupId>
            <artifactId>taverna-activity-palette-api</artifactId>
            <version>${taverna.workbench.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.taverna.workbench</groupId>
            <artifactId>taverna-menu-api</artifactId>
            <version>${taverna.workbench.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.taverna.workbench</groupId>
            <artifactId>taverna-contextual-views-api</artifactId>
            <version>${taverna.workbench.version}</version>
        </dependency>
        <dependency>
            <groupId>org.apache.taverna.workbench</groupId>
            <artifactId>taverna-activity-tools</artifactId>
            <version>${taverna.workbench.version}</version>
        </dependency>
```

stain commented on taverna-cwl-activity-ui/pom.xml:
```diff
@@ -65,27 +60,45 @@
 			<groupId>net.sf.taverna.t2.taverna-workbench</groupId>
 			<artifactId>workbench-dev</artifactId>
 			<version>${t2.workbench.version}</version>
-			<scope>test</scope>
```
`workbench-dev` must be commented out as it does not exist in Taverna 3 :-(  -- and anyway it should stay in `<scope>test`

stain commented on taverna-cwl-activity-ui/pom.xml:
```diff
@@ -178,6 +191,7 @@
 
 	<!-- Added these properties to configure new dependencies -->
 	<properties>
+		<t2.ui.impl.version>1.5</t2.ui.impl.version>
```
The `t2.*` properties should not be used in Taverna 3.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceDesc.java:
```diff
@@ -16,79 +16,67 @@
 
+import com.fasterxml.jackson.databind.JsonNode;
+
 import net.sf.taverna.t2.servicedescriptions.ServiceDescription;
```
These `import`s need to be updated to their `org.apache.taverna` equivalents.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceDesc.java:
```diff
@@ -16,79 +16,67 @@
 import net.sf.taverna.t2.workflowmodel.processor.activity.Activity;
 
-public class CwlServiceDesc extends ServiceDescription<CwlActivityConfigurationBean> {
+public class CwlServiceDesc extends ServiceDescription<JsonNode > {
```
+1 for the move to `JsonNode` - except in Taverna 3 you should not need the `<JsonNode>` generics - rather the configuration is always an [`Configuration`](https://taverna.incubator.apache.org/javadoc/taverna-language/org/apache/taverna/scufl2/api/configurations/Configuration.html) object which has `getJson*()` methods.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceDesc.java:
```diff
@@ -16,79 +16,67 @@
-		configurationBean.setCwlConfigurations(cwlConfiguration);
-		configurationBean.setToolName(toolName);
-		return configurationBean;
+	public Class<? extends Activity<JsonNode>> getActivityClass() {
```
This should be replaced in Taverna 3 with the [`getActivityType()`](https://github.com/apache/incubator-taverna-workbench/blob/master/taverna-activity-palette-api/src/main/java/org/apache/taverna/servicedescriptions/ServiceDescription.java#L45) method returning an URI which must match the `CWLActivityFactory` URI.   (This URI is used to match up the corresponding `ActivityFactory`)

I would suggest an URI like https://taverna.apache.org/ns/2016/activity/cwl for both.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceDesc.java:
```diff
@@ -16,79 +16,67 @@
-
-	public Map getCwlConfiguration() {
+	@Override
+	public JsonNode getActivityConfiguration() {
```
+1 - except you will need to wrap it in a `Configuration` object like in [super.getActivityConfiguration](https://github.com/apache/incubator-taverna-workbench/blob/master/taverna-activity-palette-api/src/main/java/org/apache/taverna/servicedescriptions/ServiceDescription.java#L48) - so perhaps this would work:

``` java
@Override
public Configuration getActivityConfiguration() {
  Configuration c = super.getActivityConfiguration();
  c.setJson(cwlConfiguration);
  return c;
}
```

(alternative use `c.getJsonAsObjectNode()` and manipulate it)

The _configuration type_ (as opposed to the above _activity type_ can be the URI https://taverna.apache.org/ns/2016/activity/cwl#Config - e.g. the default from the superclass.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceProvider.java:
```diff
@@ -16,72 +16,103 @@
+				try {
+					cwlFile = (Map) reader.load(new FileInputStream(path.toFile()));
+					JsonNode config = createJsonNode(p,cwlFile);
+					// Creating CWl service Description
```
Great - this should load the Yaml in parallell - assuming `getYamlReader()` is thread-safe?

Instead of `new FileInputStream(path.toFile())` you can do `Files.newInputStream(path)` - which should be wrapped in a (try...) resources block to ensure the input stream is closed.  (e.g. `try (InputStream is = Files.newInputStream(path)`).

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/Testing.java:
```diff
@@ -18,141 +18,63 @@
+import com.fasterxml.jackson.databind.ObjectWriter;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
 public class Testing {
```
If this class is for testing, can we move it to same package, but under `src/test/java` ? 

I'm not sure what it is testing.. is it more for experimenting?  If it can be rewritten to a junit test with `@Test` methods, that would be ideal.

stain commented on taverna-cwl-activity/src/main/java/org/apache/taverna/cwl/CwlActivityFactory.java:
```diff
@@ -0,0 +1,102 @@
+
+	@Override
+	public URI getActivityType() {
+		return null;
```
The `getActivityType()` can't be null, this is used to match up the factory from the SCUFL2 workflow and when adding from the service provider. So as mentioned above, perhaps just return `URI.create("https://taverna.apache.org/ns/2016/activity/cwl")` ?

stain commented on taverna-cwl-activity/src/main/java/org/apache/taverna/cwl/CwlActivityFactory.java:
```diff
@@ -0,0 +1,102 @@
+
+	@Override
+	public JsonNode getActivityConfigurationSchema() {
+		return null;
```
I'm not sure if this would pass with `null`. An example minimal schema which you could potentially adapt:

https://github.com/apache/incubator-taverna-plugin-gis/blob/master/taverna-gis-activity/src/main/resources/schema.json

returned as:

https://github.com/apache/incubator-taverna-plugin-gis/blob/master/taverna-gis-activity/src/main/java/org/apache/taverna/gis/GisActivityFactory.java#L55

``` java
    @Override
    public JsonNode getActivityConfigurationSchema() {
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            return objectMapper.readTree(getClass().getResource("/schema.json"));
        } catch (IOException e) {
            return objectMapper.createObjectNode();
        }
}
```

stain commented on taverna-cwl-activity/src/main/java/org/apache/taverna/cwl/CwlActivityFactory.java:
```diff
@@ -0,0 +1,102 @@
+	}
+
+	@Override
+	public Set<ActivityInputPort> getInputPorts(JsonNode configuration) throws ActivityConfigurationException {
```
Yes, great! This also uses CWLUtil, so that is nice.

stain commented on taverna-cwl-activity/src/main/java/org/apache/taverna/cwl/CwlActivityFactory.java:
```diff
@@ -0,0 +1,102 @@
+			
+	
+			//get the processed data
+			HashMap<String, Integer>  processedInputs= cwlUtil.processInputDepths();
```
It's not common for API methods in Java to be so specific about which implementation they return, but rather use the interface. Perhaps change `processInputDepths()` and friends to return/accept `Map<String, Integer>`  rather than `HashMap`?

Similar for `ArrayList` -> `List`.

stain commented on taverna-cwl-activity/src/main/resources/META-INF/spring/cwl-activity-context.xml:
```diff
@@ -0,0 +1,29 @@
+                      http://www.springframework.org/schema/beans/spring-beans.xsd">
+
+
+	<bean id="cwlActivityFactory" class="org.apache.taverna.cwl.CwlActivityFactory">
```
+1, this looks correct to me

stain commented on taverna-cwl-utilities/CWLFiles/customtool1.cwl:
```diff
@@ -0,0 +1,67 @@
+################################################################################
```
Could we move `CWLFiles/` to `src/test/resources` to avoid dependency on correct relative paths?  (E.g. so it works in both the IDE and Maven).

Then within the test you can try to use `getClass().getResource()` to retrieve it.. or perhaps that is awkward for methods that need a `Path`?

stain commented on taverna-cwl-utilities/src/main/java/org/apache/taverna/cwl/utilities/CWLUtil.java:
```diff
@@ -0,0 +1,216 @@
+package org.apache.taverna.cwl.utilities;
```
Could you add the ASF header..? Thanks! :-)

stain commented on taverna-cwl-utilities/src/main/java/org/apache/taverna/cwl/utilities/CWLUtil.java:
```diff
@@ -0,0 +1,216 @@
+		return processdetails(cwlFile.get(OUTPUTS));
+	}
+
+	public HashMap<String, Integer> process(JsonNode inputs) {
```
It is unclear what this public `process()` method does or returns... could you add a `/**` javadocs explaining inputs and outputs for these methods?

Also as mentioned before, `HashMap` -> `Map` :-)

stain commented on taverna-cwl-utilities/src/test/java/org/apache/taverna/cwl/utilities/CWLUtilTest.java:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	@Test
+	public void processTest() {
```
Great!

stain commented:
Thank you for showing your progress!

I've left lots of comments under "Files changed" in the  pull request - don't be discouraged! Sorry I hadn't reviewed many of the files in this detail before.

As for the Taverna 3 upgrade your activity is almost there - but the activity-ui is still using the old imports, which means it's now kind of "in between". :-)  I think once you get the right `<dependency>` and `import`s added it would show up better in Eclipse what needs to be changed - but I guess you will have more questions after that.

Just yesterday I helped our local student @edikaradumi in updating https://github.com/apache/incubator-taverna-plugin-gis/tree/master/taverna-gis-activity-ui/src/main/java/org/apache/taverna/gis/ui/serviceprovider to Taverna 3 - I must admit we didn't get to tested it in the T3 workbench yet, but  it should compile. :)

ThilinaManamgoda committed:
resolve T# dependecies

ThilinaManamgoda committed:
removed Testing.java

ThilinaManamgoda committed:
moved CWLFiles Folder

ThilinaManamgoda committed:
1. add Icon resource folder to ui module
2. Service provider is configured using Configuration object
3. new CWL Activity Factory is added
4. Schema.json file is added which defines the structure of the CWL activity configuration

ThilinaManamgoda committed:
Java Doc

ThilinaManamgoda committed:
Java Doc

ThilinaManamgoda committed:
Fixme setuped

ThilinaManamgoda committed:
javadocs

ThilinaManamgoda committed:
unit tests

ThilinaManamgoda committed:
activity problem

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceProvider.java:
```diff
@@ -16,72 +16,167 @@
+			return defaultPaths;
+		}
+
+		return (ArrayList<Path>) Arrays.asList(Paths.get(userInput));
```
You should not assume `Arrays.asList()` return an `ArrayList`! This could break with any JDK patch update.

stain commented on taverna-cwl-activity-ui/src/main/java/org/apache/taverna/cwl/ui/serviceprovider/CwlServiceProvider.java:
```diff
@@ -16,72 +16,167 @@
+	}
 
-	
+	private ArrayList<Path> getPath() {
```
Return as interface `List<Path>` instead of hard-coding `ArrayList` implementation - we don't need ArrayList-specific features.

stain commented on taverna-cwl-activity/src/main/resources/CWLFiles/customtoolrecord.cwl:
```diff
@@ -0,0 +1,41 @@
+################################################################################
```
Could test resources be in `src/test/resources`

ThilinaManamgoda committed:
moved CWLFiles content to src/test/resources
activity and activity-ui share common test reources

ThilinaManamgoda committed:
moved all the process done in the cwlContextualview to utilities module
unit tests

ThilinaManamgoda committed:
contextual view code review

ThilinaManamgoda committed:
removed contextualview test

ThilinaManamgoda committed:
new array type

ThilinaManamgoda committed:
add unit test for CWLUtil

ThilinaManamgoda committed:
removed log file

ThilinaManamgoda committed:
cwlDumyActivity.class to CwlDummyActivity.class

ThilinaManamgoda committed:
upgradede for CWL v1.0

ThilinaManamgoda committed:
can process int[]

ThilinaManamgoda committed:
cwl v1.0 unitTest

ThilinaManamgoda committed:
unit test : json files

ThilinaManamgoda committed:
cwlutil => cwlUtil

ThilinaManamgoda committed:
testresource class constructor

ThilinaManamgoda committed:
organized imports
